### PR TITLE
[SYCL][Test] Devicelib test

### PR DESCRIPTION
**Pull request by haonanya**
1. add test for all math function.
2. In llvm/sycl/source/detail/devicelib/CMakeLists.txt, libsycl-cmath.o depends on device_math.h,
modify here.

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @bader , @asavonic , @andykaylor 
We plan to upload more tests for cmath device library, could you help review?
Thank you very much.

File: sycl/test/devicelib/math_test.cpp [hunk 1]
```diff
@@ -9,64 +9,821 @@ namespace s = cl::sycl;
...
       });
     });
   }
 
-  assert(result_f == 0);
+  assert(std::isnan(result));
 }
 
 void device_math_test(s::queue &deviceQueue) {
   device_cos_test(deviceQueue);
```
Comment (bader) [inline sycl/test/devicelib/math_test.cpp]: This will be terribly slow.
We need to make LIT tests as fast possible.
Also each test case duplicates a ton of a boilerplate code.
First idea:
- put all math function calls into a single kernel.

File: sycl/test/devicelib/math_test.cpp [hunk 2]
```diff
@@ -9,64 +9,821 @@ namespace s = cl::sycl;
...
+      auto quo_access = buffer2.template get_access<sycl_write>(cgh);
+      cgh.single_task<class DeviceRemquo>([=]() {
+        float a = 0.5;
+        float b = 1;
+        res_access[0] = remquof(a, b, &quo_access[0]);
+      });
+    });
+  }
+
+  assert(result == ref);
```
Comment (bader) [inline sycl/test/devicelib/math_test.cpp]: Is this reliable way to validate math functions result?
NOTE: math functions are allowed to produce "inaccurate" results.

File: sycl/test/devicelib/math_test.cpp [hunk 3]
```diff
@@ -4,69 +4,217 @@
...
+      cgh.single_task<class DeviceMathTest>([=]() {
+        int i = 0;
+        {
+          float a = 0;
+          res_access[i++] = cosf(a);
+        }
+        {
+          float a = 0;
+          res_access[i++] = sinf(a);
+        }
```
Comment (bader) [inline sycl/test/devicelib/math_test.cpp]: Note this code is 90% repeats itself.
Only input value and function name are different.
Can use something like:
```suggestion
          res_access[i++] = cosf(0.f);
          res_access[i++] = sinf(0.f);
```
?

File: sycl/test/devicelib/math_utils.hpp [hunk 4]
```diff
@@ -1,23 +1,28 @@
...
   else {
     if (x != 0 && y != 0) {
-      T max_v = fmax(abs(x), abs(y));
-      return (abs(x - y) / max_v) <
-              numeric_limits<T>::epsilon() * 100;
-    }
-    else {
+      T max_v = std::fmax(std::abs(x), std::abs(y));
+      return (std::abs(x - y) / max_v) <
+             std::numeric_limits<T>::epsilon() * 100;
```
[the remark this answers is not in the document]
Comment (jinge90) [reply sycl/test/devicelib/math_utils.hpp]: Hi, Alexey.
In fact, we added those cases only to test basic functionality of math functions provided by device library. The accuracy is not meant to be covered by these tests, all those input values of cmath functions are very simple ones.  In fact, we are doing some accuracy tests on device library's <cmath> and <complex>, all input will be much more complicated and different threshold are used and compared.
Thank you very much.

File: sycl/test/devicelib/math_utils.hpp [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
-  if (x == y)
-    return true;
+  bool ret;
+
+  // At least one input is nan
+  if (std::isnan(x) || std::isnan(y))
+    ret = std::isnan(x) && std::isnan(y);
+
+  // At least one input is inf
+  else if (std::isinf(x) || std::isinf(y))
```
Comment (bader) [inline sycl/test/devicelib/math_utils.hpp]: I found the code with early exit more readable than this version.
```suggestion
  // At least one input is nan
  if (std::isnan(x) || std::isnan(y))
    return std::isnan(x) && std::isnan(y);

  // At least one input is inf
  if (std::isinf(x) || std::isinf(y))
```

**Comment by bader** [comment]
Please, fix code formatting issues.

File: sycl/test/devicelib/math_utils.hpp (hunk 5 above)
Comment (bader) [reply sycl/test/devicelib/math_utils.hpp]: Fix for fixing this, but could you apply this comment to the file, please?

Comment (haonanya) [reply sycl/test/devicelib/math_utils.hpp]: Hi, Alexey.
I am sorry to make things complicated. I am new user.
Thanks for your patience.

File: sycl/test/devicelib/math_utils.hpp [hunk 6]
```diff
@@ -1,25 +1,33 @@
 #ifndef MATH_UTILS
 #include <complex>
 #include <limits>
-using namespace std;
 // T must be float-point type
 template <typename T>
 bool is_about_FP(T x, T y) {
```
Comment (keryell) [inline sycl/test/devicelib/math_utils.hpp]: While you are here, could you add a comment explaining what this function... is_about? :-)

File: sycl/test/devicelib/math_utils.hpp [hunk 7]
```diff
@@ -1,25 +1,33 @@
...
   else {
+    T threshold = std::numeric_limits<T>::epsilon() * 100;
     if (x != 0 && y != 0) {
-      T max_v = fmax(abs(x), abs(y));
-      return (abs(x - y) / max_v) <
-              numeric_limits<T>::epsilon() * 100;
-    }
-    else {
+      T max_v = std::fmax(std::abs(x), std::abs(y));
+      ret = (std::abs(x - y) / max_v) < threshold;
```
Comment (keryell) [inline sycl/test/devicelib/math_utils.hpp]:
```suggestion
      return std::abs(x - y) < threshold*max_v;
```
is more efficient and probably more in the spirit of what this function is_about.

File: sycl/test/devicelib/math_utils.hpp (hunk 6 above)
Comment (haonanya) [reply sycl/test/devicelib/math_utils.hpp]: Hi, Ronan.
Since it is not proper to compare float point using operator ==, this function measure whether the result of cmath function from kernel is close to the reference and machine epsilon is used as threshold in this function.
Thanks very much.

Comment (keryell) [reply sycl/test/devicelib/math_utils.hpp]: Ah quite clearer with a comment! :-)

Then perhaps a clearer name for the function would help...

`fuzzy_equal_fp` ? `approx_equal_fp` ? I am not an native English speaker so I am not very helpful here...

File: sycl/test/devicelib/math_utils.hpp [hunk 8]
```diff
@@ -1,25 +1,37 @@
...
+    } else {
       if (x != 0)
-        return abs(x) < numeric_limits<T>::epsilon() * 100;
+        ret = std::abs(x) < threshold;
       else
-        return abs(y) < numeric_limits<T>::epsilon() * 100;
+        ret = std::abs(y) < threshold;
     }
   }
+  return ret;
```
Comment (keryell) [inline sycl/test/devicelib/math_utils.hpp]: In this kind of code it is probably clearer to remove `ret` and just have as many early return as required.

Comment (haonanya) [reply sycl/test/devicelib/math_utils.hpp]: Thank for your comment, early return is applied to new commit and approx_equal_fp is used as the name of float point comparison.

**Comment by bader** [comment]
@haonanya, please, resolve conflicts.

File: sycl/test/devicelib/math_utils.hpp [hunk 9]
```diff
@@ -1,23 +1,33 @@
...
+  // At least one input is nan
+  if (std::isnan(x) || std::isnan(y))
+    return std::isnan(x) && std::isnan(y);
+
+  // At least one input is inf
+  if (std::isinf(x) || std::isinf(y))
+    return (x == y);
+
+  // two finite
   else {
```
Comment (bader) [inline sycl/test/devicelib/math_utils.hpp:21]: Please, [don't use else after return](https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return).
Apply here and all the cases below.

**Comment by bader** [comment]
LGTM, but I'd like to have reviews from @asavonic and @romanovvlad.
Thanks!

File: sycl/test/devicelib/cmath_fp64_test.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
+// UNSUPPORTED: windows
+// RUN: %clangxx -fsycl -c %s -o %t.o
+// RUN: %clangxx -fsycl %t.o %sycl_libs_dir/libsycl-cmath-fp64.o -o %t.out
```
Comment (asavonic) [inline sycl/test/devicelib/cmath_fp64_test.cpp:3]: Please add a RUN line for execution here and in all other tests.
Otherwise this just tests for compilation and no runtime validation is performed.

Comment (andykaylor) [reply sycl/test/devicelib/cmath_fp64_test.cpp:3]: If these are going to run the compiled program, they should be moved to end-to-end testing.

Comment (jinge90) [reply sycl/test/devicelib/cmath_fp64_test.cpp:3]: Hi, @andykaylor  and @asavonic 
Could we enable the run command for cmath tests on CPU device currently?
For GPU, only OCL RT can work. For CPU, all cmath and complex tests should work in theory but OCL CPU RT has an issue which will lead to crash on some platforms, the fix has not been latest OCL CPU RT, so we may encounter some unexpected crash if enable complex tests.
Thank you very much.

Comment (asavonic) [reply sycl/test/devicelib/cmath_fp64_test.cpp:3]: @andykaylor 
> If these are going to run the compiled program, they should be moved to end-to-end testing.

Most of the tests in `sycl/test` directory have RUN lines to test on available devices.
This test in particular has a lot of runtime asserts.

> Could we enable the run command for cmath tests on CPU device currently?

I think they should be enabled for all devices you expect this code to work. I assume it should work for CPU and GPU, right? FPGA is probably not going  work, because AOT compilation is not support for device libraries yet.

Comment (jinge90) [reply sycl/test/devicelib/cmath_fp64_test.cpp:3]: Yes, the tests are expected to work on both CPU and GPU. However, for GPU, we need to force the tests to run on OCL GPU runtime as L0 hasn't supported device library. For CPU, if we enable complex tests, crash may happen on some platform due to OCL CPU runtime issues. So, for GPU, can we add export SYCL_BE=PI_OPENCL?

Comment (asavonic) [reply sycl/test/devicelib/cmath_fp64_test.cpp:3]: If a test doesn't work on some platform due to a bug, it should be disabled temporarily.

**Comment by bader** [comment]
@asavonic, ping.
